backend/src/email/middlewares/email-transactional.middleware.ts

```diff
@@ -23,53 +26,120 @@ export const InitEmailTransactionalMiddleware = (
     next: NextFunction
   ): Promise<void> {
-    const ACTION = 'sendMessage'
+    const action = 'saveMessage'
+    logger.info({ message: 'Saving email', action })
+    const {
+      subject,
+      body,
+      from,
+      recipient,
+      reply_to: replyTo,
+      attachments,
+    }: ReqBody = req.body
 
-    try {
-      const {
+    const { fromName, fromAddress } = parseFromAddress(from)
+    const userEmail = (await authService.findUser(req.session?.user?.id))?.email
+    const emailMessageTx = await EmailMessageTx.create({
+      userEmail,
+      fromName,
+      fromAddress,
+      recipient,
+      params: {
         subject,
         body,
         from,
-        recipient,
-        reply_to: replyTo,
-        attachments,
-      } = req.body
+        replyTo,
+      },
+      hasAttachment: !!attachments,
+      attachmentS3Object: null,
+      status: TransactionalEmailMessageStatus.Unsent,
+      sentAt: null,
+      errorCode: null,
+      // not sure why unknown is needed to silence TS (yet other parts of the code base can just use `as Model` directly hmm)
+    } as unknown as EmailMessageTx)
+    if (!emailMessageTx) {
+      // TODO: log error, etc.
+      res.status(500).json({
+        message: 'Unable to create entry in email_message_tx',
+      })
+      throw new Error('Unable to create entry in email_message_tx')
+    }
+    // attach to body for subsequent middlewares
+    req.body.userEmail = userEmail
+    req.body.emailMessageTxId = emailMessageTx.id
+    next()
+  }
 
-      logger.info({ message: 'Sending email', action: ACTION })
+  async function sendMessage(
+    req: Request,
+    res: Response,
+    next: NextFunction
+  ): Promise<void> {
+    const action = 'sendMessage'
+    logger.info({ message: 'Sending email', action })
+    const {
+      subject,
+      body,
+      from,
+      recipient,
+      reply_to: replyTo,
+      attachments,
+      userEmail, // added by saveMessage; avoid unnecessary DB query
+      emailMessageTxId, // added by saveMessage
+    }: ReqBody & { userEmail: string; emailMessageTxId: number } = req.body
+
+    try {
       await EmailTransactionalService.sendMessage({
         subject,
         body,
         from,
         recipient,
-        replyTo:
-          replyTo ?? (await authService.findUser(req.session?.user?.id))?.email,
+        replyTo: replyTo ?? userEmail,
         attachments,
+        emailMessageTxId,
       })
+      await EmailMessageTx.update(
+        {
+          status: TransactionalEmailMessageStatus.Accepted,
+          sentAt: new Date(),
+        },
+        {
+          where: { id: emailMessageTxId },
+        }
+      )
       res.sendStatus(202)
-    } catch (err) {
+    } catch (error) {
       logger.error({
         message: 'Failed to send email',
-        action: ACTION,
-        error: err,
+        action,
+        error,
       })
 
       const BAD_REQUEST_ERRORS = [
-        TemplateError,
+        InvalidMessageError,
```

Comment:
thought it's weird to use `TemplateError` in API call → have created a new error under `transactional.error.ts` instead